feat: 🔥 native global search (client side) 🔥

vpicone opened the pull request:
## Summary:

This adds global search functionality to every site that enables it. RIP Algolia.

### Changed

- `gatsby-node.js` link Mdx nodes with their corresponding pages nodes. This allows us to index Mdx body content along with the page data.
- `GlobalSearchInput` – rewrite to query index fields, remove useSearch hook

### New
- Use [lunr](https://lunrjs.com/) to index and query page data.

	The indexing algorithm favors `title` and `keywords`, but queries pages based on the following fields: `title`, `keywords`, `description`, `content`, and `path`.
- Add styles to bring the search in line with the other shell components
- Add descriptions to each page to populate the list items


### Todo

- [x] arrow key navigation
- [x] remove old search styles in favor of modules
- [x] expose resolvers as theme option, cleanup resolver logic
- [x] mobile

abbeyhrt commented:
This might be out of scope, but do the search results need to be keyboard accessible with up and down arrows like the other dropdown menus in Carbon?

vpicone commented:
@abbeyhrt definitely not out of scope! It's in my to-do list at the top of the PR.

vpicone commented on packages/gatsby-theme-carbon/src/components/GlobalSearch/GlobalSearchInput.js:
```diff
@@ -1,60 +1,112 @@
     <div
       ref={searchRef}
-      className="bx--search bx--search--lg bx--search--dark"
+      className={`bx--search bx--search--lg bx--search--dark ${container}`}
```
There's enough overrides here I think I'm just going to not use the carbon search styles for now. We're not using the React Search component here so I don't feel like keeping the native carbons styles is a priority.

vpicone commented:
@emyarod sorry their for this PR, at least while we wait for design signoff/spec

jeanservaas commented:
qq, is there a design spec for this that we're referring to? Is it meant to be the same spec as product search? Usually these are attached to an issue, just want a quick way to compare it visually without having to dig around. 

- Also the close 'x' jumps a bit on hover

vpicone commented:
@jeanservaas the default search component doesn't include the autocomplete window. This component is a mash up of the search component and a combobox.

I made it dark to be closer in line with our menu drop downs in shell as well as the switcher: http://react.carbondesignsystem.com/?path=/story/ui-shell--header-base-w-navigation

Also the X button is doing that in the master branch right now, this doesn't touch that bit.

vpicone commented:
@jeanservaas there's no design spec that I'm aware of, but there's different data presented and we're not beholden to the Algolia styling limitations any more. I think that's why @designertyler mentioned an opportunity to revisit the design of this element.

jeanservaas commented:
OK one other qq about focus state behavior, is it functioning as expected to have both focus states (close button focus and field focus) triggering at the same time.

<img width="806" alt="image" src="https://user-images.githubusercontent.com/43144260/64032291-bf8abf80-cb0f-11e9-84fc-49775218c567.png">


vs. something like this:

<img width="321" alt="image" src="https://user-images.githubusercontent.com/43144260/64032322-d29d8f80-cb0f-11e9-9c7d-801730fb8ab7.png">


I'm just making sure that it's accessible.

vpicone commented:
@jeanservaas not sure, that's also something that was in the old carbon website and is in the current implementation, the main modification here is to the list that pops up.

designertyler commented:
Design specs for website search
https://ibm.box.com/s/4e7tu2m6qyz6iiumrmrcqxykq8askh8x

alisonjoseph commented:
Looking good!! Not sure how I feel about the animation, seems a little jerky/bouncy? when you close the search? tagging @shixiedesign our animation expert to take a look.

emyarod commented:
the search input animation might be related to https://github.com/carbon-design-system/carbon/issues/2339

vpicone commented:
@alisonjoseph double check you're using the latest deployment, the latest push smoothed it out a bunch. The most sudden bit is moving the focus, I could try delaying that though.

This is slowed down 10x:

![slow](https://user-images.githubusercontent.com/4078018/64467845-b470fa00-d0e2-11e9-8fce-21f695f891f6.gif)

alisonjoseph commented:
I think its the moving of the search icon thats weird for me, but if @shixiedesign is ok with it then I am too. 😄

shixiedesign commented:
Amazing work! Great application of the easing curves and thank you for the effort. @vpicone only 2 small comments: 
1. the X flies in from the right side, over the App Switcher button. Can we have X appear instead of flying in? Or it can fly in from behind the App Switcher? 
![Sep-09-2019 17-23-09](https://user-images.githubusercontent.com/15144993/64570613-c0e59480-d326-11e9-996e-5970715b580a.gif)
2. Search button doesn't have animation onHover. Background color should change with `productive standard easing` and duration `duration--fast-01` or `70ms`. This is only noticeable because the App switcher button next to it has an animation, so it feels inconsistent when I hover around the area. See second half of above gif.

vpicone commented:
@shixiedesign I'll check the X out, thanks!

For the search button, I didn't add hover animations to the search because there is no "search" action. If the search button _was_ clickable, it'd take us to a "search results" page. But, we don't have such a page so the only actions available to a user are to 1) click a link from the list or 2) to press the clear/close button.

vpicone commented:
New from yesterday 9/10:
- search enabled by default
- added/updated documentation

fixed:
- Search button is only focused on close if we took over focus management in the first place.
- search and title text alignment @emyarod 
- close button animation layering @shixiedesign 
- cleanup old shadowing hook
- various ui shell active/focus state fixes
- fix #406

vpicone commented:
@emyarod I moved the managing focus state to context, should be good to go now
@shixiedesign switcher hover background is fixed

vpicone commented:
@emyarod @jeanservaas went ham and merged it, do you mind making a new issue that describes the behavior/expected behavior?

emyarod commented:
@vpicone sure, opened https://github.com/carbon-design-system/gatsby-theme-carbon/issues/412
